typesetters/base.lua

```diff
@@ -315,18 +322,160 @@ function typesetter:breakIntoLines (nodelist, breakWidth)
+        break
+      end
+      if n.is_glue then hasGlue = true end
+    end
+  end
+  return last, hasGlue
+end
+local function getFirstShape(nodelist)
+  local first
+  local hasGlue
+  if nodelist then
+    -- The node list may contain nnodes, penalties, kern and glue
+    -- We skip the latter, and retrieve the first shaped item.
+    for i = 1, #nodelist do
+      local n = nodelist[i]
+      if n.is_nnode then
+        local items = n.nodes[1].value.items
+        first = items[1]
+        break
+      end
+      if n.is_kern and n.subtype == "punctspace" then
+        -- Some languages such as French insert a special space around
+        -- punctuations. In those case, we should not need italic correction.
+        break
+      end
+      if n.is_glue then hasGlue = true end
+    end
+  end
+  return first, hasGlue
+end
+
+local function fromItalicCorrection (precShape, curShape)
+  local xOffset
+  if not curShape or not precShape then
+    xOffset = 0
+  else
+    -- Computing italic correction is at best heuristics.
+    -- The strong assumption is that italic is slanted to the right.
+    -- Thus, the part of the character that goes beyond its width is usually
+    -- maximal at the top of the glyph.
+    -- E.g. consider a "f", that would be the top hook extent.
+    -- Pathological cases exist, such as fonts with a Q with a long tail,
+    -- but these will rarely occur in usual languages. For instance, Klingon's
+    -- "QaQ" might be an issue, but there's not much we can do...
+    -- Another assumption is that we can distribute that extent in proportion
+    -- with the next character's height.
+    -- This might not work that well with non-Latin scripts.
+    local d = precShape.glyphWidth + precShape.x_bearing
+    local delta = d > precShape.width and d - precShape.width or 0
+    xOffset = precShape.height <= curShape.height
+      and delta
+      or delta * curShape.height / precShape.height
+  end
+  return xOffset
+end
+
+local function toItalicCorrection (precShape, curShape)
+  if not SILE.settings:get("typesetter.italicCorrection") then return end
+  local xOffset
+  if not curShape or not precShape then
+    xOffset = 0
+  else
+    -- Same assumptions as fromItalicCorrection(), but on the starting side of
+    -- the glyph.
+    local d = curShape.x_bearing
+    local delta = d < 0 and -d or 0
+    xOffset = precShape.depth >= curShape.depth
+      and delta
+      or delta * precShape.depth / curShape.depth
+  end
+  return xOffset
+end
+
+local function isItalicLike(nnode)
+  -- We could do...
+  --  return nnode and string.lower(nnode.options.style) == "italic"
+  -- But it's probably more robust to use the italic angle, so that
+  -- thin italic, oblique or slanted fonts etc. may work too.
+  local ot = require("core.opentype-parser")
+  local face = SILE.font.cache(nnode.options, SILE.shaper.getFace)
+  local font = ot.parseFont(face)
+  return font.post.italicAngle ~= 0
```

Comment:
Reading https://learn.microsoft.com/en-us/typography/opentype/spec/post#post-table-and-opentype-font-variations:
> Note: The italicAngle value is not adjusted by variation data since this corresponds to the 'slnt' variation axis that can be used to define a font’s variation space. Appropriate post.italicAngle values for a variation instance can be derived from the 'slnt' user coordinates that are used to select a particular variation instance. See the discussion of the 'slnt' axis in the [OpenType Design-Variation Axis Tag Registry](https://learn.microsoft.com/en-us/typography/opentype/spec/dvaraxisreg) for details on the relationship between italicAngle and the 'slnt' axis.

I followed the link, but it's all beyond my knowledge.
Would we possibly need to make this italic detection more robust, and how?

Reply:
I think we will need to make this more robust, but I also don't think there is any harm in starting with this.

The two situations that don't seem to be covered are:

1. Incorrectly built fonts that *deserve* a value in this data and were not given one.
2. Variable fonts.

Rolling with this should catch *correctly built* static font instances, which is not a bad chunk to bite off first.